### ISSUE TITLE
MNIST example 2x slower on GPU

### ISSUE THREAD

**Issue by dhdaines** [fchollet/keras#249]
The MNIST example seems to be running about 2x slower for me on a GPU (Tesla K10.G1.8GB) than on CPU.

When I run it with profiling it seems that it is spending a large amount of time in GpuFromHost.  This would indicate that the data is not being loaded onto the GPU with shared variables (and in fact I can't really tell where in the Keras code this would be taking place) and thus each minibatch must be transferred to the GPU at each iteration of training.

Yet all the other examples give instructions for running on GPU, and thus Keras is presumably optimized for GPU. Am I missing something here?

**Comment by fchollet** [comment]
> This would indicate that the data is not being loaded onto the GPU with shared variables (and in fact I can't really tell where in the Keras code this would be taking place) and thus each minibatch must be transferred to the GPU at each iteration of training.

Data is being loaded on the GPU one batch at a time, as it should be. Loading all the data at once would only be viable for very small problems (like MNIST) and would crash your GPU most of the time. 

I just upped the batch_size from 64 to 128. The example now runs 2x faster on GPU compared to CPU, and accuracy is slightly improved as well.

**Comment by entron** [comment]
I tried use keras with imagenet. The speed is about 170 times slower than using just theano. Probably the only difference is copy per batch or per chunk.

**Comment by fchollet** [comment]
Then adjust the batch size.

**Comment by tleeuwenburg** [comment]
Maybe it would be worth including some brief notes on performance
optimisation somewhere in the docs.

On 13 August 2015 at 13:41, François Chollet <notifications@github.com>
wrote:

> Then adjust the batch size.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/249#issuecomment-130520899>.
>



-- 
--------------------------------------------------
Tennessee Leeuwenburg
http://myownhat.blogspot.com/
"Don't believe everything you think"

**Comment by fchollet** [comment]
Yes, I think we ned a FAQ to answer these questions, as well as many others.

**Comment by entron** [comment]
I am already use batch_size = 128. If I increase it too much then the GPU memory (4GB) is not enough to hold the model. This is the model I use. I am still checking whether I made some error somewhere to cause such drastic speed difference.

```
def ImageNet(nb_classes):
    model = Sequential()

    model.add(Convolution2D(64, 3, 11, 11, subsample=(4, 4)))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(poolsize=(3, 3), stride=(2, 2)))

    model.add(Convolution2D(96, 64, 5, 5, border_mode='same'))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(poolsize=(3, 3), stride=(2, 2)))

    model.add(Convolution2D(128, 96, 3, 3, border_mode='same'))
    model.add(Activation('relu'))

    model.add(Convolution2D(256, 128, 3, 3, border_mode='same'))
    model.add(Activation('relu'))

    model.add(Convolution2D(256, 256, 3, 3, border_mode='same'))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(poolsize=(3, 3), stride=(2, 2)))

    model.add(Flatten())
    model.add(Dense(256 * 6 * 6, 4096))
    model.add(Activation('relu'))
    model.add(Dropout(0.5))

    model.add(Dense(4096, 4096))
    model.add(Activation('relu'))
    model.add(Dropout(0.5))

    model.add(Dense(4096, nb_classes))
    model.add(Activation('softmax'))
    return model
```

**Comment by entron** [comment]
I did profile and this is part of the result:

```
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  99.7%    99.7%     192.067s       9.60e+01s     C        2       1   theano.tensor.nnet.ConvGrad3D.ConvGrad3D
   0.1%    99.8%       0.155s       1.93e-02s     C        8       4   theano.sandbox.cuda.basic_ops.HostFromGpu
   0.1%    99.8%       0.110s       1.10e-02s     C       10       5   theano.sandbox.cuda.dnn.GpuDnnConv
   0.1%    99.9%       0.100s       1.25e-02s     C        8       4   theano.sandbox.cuda.dnn.GpuDnnConvGradI
   0.0%    99.9%       0.091s       1.13e-02s     C        8       4   theano.sandbox.cuda.dnn.GpuDnnConvGradW
   0.0%   100.0%       0.038s       3.01e-04s     C      126      63   theano.sandbox.cuda.basic_ops.GpuElemwise
   0.0%   100.0%       0.025s       2.53e-03s     C       10       5   theano.sandbox.cuda.basic_ops.GpuFromHost
   0.0%   100.0%       0.014s       1.76e-03s     C        8       4   theano.sandbox.cuda.blas.GpuDot22
   0.0%   100.0%       0.011s       1.83e-03s     C        6       3   theano.sandbox.cuda.dnn.GpuDnnPoolGrad
   0.0%   100.0%       0.010s       1.61e-03s     C        6       3   theano.sandbox.cuda.blas.GpuGemm
   0.0%   100.0%       0.006s       9.69e-04s     C        6       3   theano.sandbox.cuda.dnn.GpuDnnPool
```

ConvGrad3D takes too much time. I found the solution [here](http://theano-users.narkive.com/Twhmp86s/cnn-with-cudnn-order-of-magnitude-slower-than-should-be).

> the problem is that this op is being used to implement the
> gradient of a strided convolution: theano.tensor.nnet.ConvGrad3D.ConvGrad3D
> 
> It runs on the CPU, so it's very slow (as you can see in the profile). To
> work around this, you can use theano.sandbox.cuda.dnn.dnn_conv instead of
> theano.tensor.nnet.conv.conv2d in your code

I change the part of the code in keras, and now it is 170 times faster!

**Comment by fchollet** [comment]
> ConvGrad3D takes too much time. I found the solution here.

Interesting discussion. Would you like to submit a PR with this change (GPU-only)?

**Comment by entron** [comment]
PR is here: https://github.com/fchollet/keras/pull/524
Note that now cudnn is required and it can only run on GPU.

**Comment by mathDR** [comment]
Was this merged?
